### ISSUE TITLE
fred (and redis-protocol) uses wrong assumption that keys are utf8

### ISSUE THREAD

**Issue by khvzak** [aembke/fred.rs#22]
Hello,
First of all thanks for the great library!

I just want to raise an issue about wrong assumption (and behaviour) that Redis keys are valid utf8 strings.
According the [Redis data types doc](https://redis.io/topics/data-types-intro)
> Redis keys are binary safe, this means that you can use any binary sequence as a key, from a string like "foo" to the content of a JPEG file. 

Although in the `redis-protocol` library the function [`redis_keyslot`](https://docs.rs/redis-protocol/3.0.2/redis_protocol/fn.redis_keyslot.html) wrongly accepts key as `&str` type.

In combination with `String::from_utf8_lossy ` conversion in https://github.com/aembke/fred.rs/blob/main/src/commands/lua.rs#L26 it produces wrong keys.

Unfortuantely this is not a single issue. In https://github.com/aembke/redis-protocol.rs/blob/main/src/utils.rs#L360 utf8 string key is indexed using byte index which not always correct because index can be not  a UTF-8 code point boundary (which generates panic).

**Comment by aembke** [comment]
Hi @khvzak,

You're absolutely right, and I completely forgot to reconcile the differences between the current `RedisKey` and underlying logic in `redis_protocol`. When I released 4.x I started fixing this issue (since prior to 4.x even `RedisKey` assumed UTF8 strings), but I never got around to fixing everything that used it like a string under the hood. 

I'm in the middle of the next major version (5.x) right now, and I'm hoping to release that in the next week or two. I'll include these changes in that release as well.

**Comment by khvzak** [comment]
sorry @aembke I didn't provide an example of panic caused by wrong bytex indexing inside a valid unicode string.
```rust
    let buf = vec![b'{', 239, 191, 189, b'}'];
    let _ = String::from_utf8(buf).map(|s| redis_protocol::redis_keyslot(&s));
```
This code will trigger panic even with the latest redis-protocol v3.1.0 😞 

Also wondering, would makes sence instead of creating two versions of the `redis_keyslot`, use a single generic form:
```rust
pub fn redis_keyslot<S: AsRef<[u8]> + ?Sized>(key: &S) -> u16 {}
```
that accepts `&str` and `&[u8]`?

**Comment by aembke** [comment]
Hi @khvzak,

Yeah you're right. I'm already in the middle of version 4.x of `redis-protocol` so I opted to do something quick and dirty for this for a few reasons. 

The new version 4 of redis-protocol uses borrowed frames to avoid any copies of the underlying data, but the ergonomics of this are pretty bad when combined with the tokio `Decoder` interface. I've spent the last week or so trying to get around the fact that callers (in this case myself with fred) pretty much have to use a self-referential struct to implement anything non-trivial such as parsing streamed frames. I really want to avoid any unsafe code in `redis-protocol`, but that might be necessary since AFAIK I can't use `owning-ref` or friends here without duplicating complicated parsing logic, so we'll see. 

To make a long story short I assumed I wouldn't find a good solution here in time for the 5.x fred release so I just did something quick and dirty in 3.x for `redis-protocol`, but you're right I didn't consider the panic case or I would have removed the `&str` implementation entirely to avoid the panic. 

I'm going to fight the compiler a bit more in 4.x to try to find a good way to do this, and if so I'll add some fuzzing tests and change things to account for what you pointed out above. If not I'll probably just yank 3.1.0 and turn 4.x into a smaller release that mostly just does what you described above. 

We'll see what happens though. Either way I'm going to time box this to this weekend, and I may end up just scraping the current 4.x logic if I have to use too much unsafe code.
